src/call_context.rs

```diff
@@ -1,68 +1,65 @@
 // Copyright (c) DUSK NETWORK. All rights reserved.
 // Licensed under the MPL 2.0 license. See LICENSE file in the project root for details.
 
-use std::marker::PhantomData;
-use std::mem;
-
-use dataview::Pod;
-use dusk_abi::H256;
-use kelvin::{ByteHash, ValRef, ValRefMut};
+use canonical::{ByteSink, ByteSource, Canon, Sink, Store};
+use dusk_abi::{Query, ReturnValue, Transaction};
 
 use wasmi::{
-    ExternVal, Externals, ImportsBuilder, MemoryRef, ModuleImportResolver,
-    ModuleInstance, RuntimeArgs, RuntimeValue, Trap, TrapKind,
+    Externals, ImportsBuilder, MemoryRef, ModuleImportResolver, RuntimeArgs,
+    RuntimeValue, Trap, TrapKind,
 };
 
-use crate::contract::MeteredContract;
+use crate::contract::ContractId;
 use crate::gas::GasMeter;
-use crate::state::{NetworkState, Storage};
+use crate::state::NetworkState;
 use crate::VMError;
 
-pub trait Resolver<H: ByteHash>:
-    Invoke<H> + ModuleImportResolver + Default + Clone
+pub trait Resolver<S: Store>:
+    Invoke<S> + ModuleImportResolver + Clone + Default
 {
 }
 
 pub use crate::resolver::CompoundResolver as StandardABI;
 
+#[derive(Debug)]
+enum Argument {
+    Query(Query),
+    Transaction(Transaction),
+}
+
 pub struct StackFrame {
-    callee: H256,
-    opcode: usize,
-    argument_ptr: usize,
-    argument_len: usize,
-    return_ptr: usize,
-    return_len: usize,
+    callee: ContractId,
+    argument: Argument,
+    ret: ReturnValue,
     memory: MemoryRef,
 }
 
 impl std::fmt::Debug for StackFrame {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(
-            f,
-            "(arg: {} return: {})",
-            self.argument_ptr, self.return_ptr
-        )
+        write!(f, "(arg: {:?} return: {:?})", self.argument, self.ret)
     }
 }
 
 impl StackFrame {
-    fn new(
-        callee: H256,
+    fn new_query(callee: &ContractId, memory: MemoryRef, query: Query) -> Self {
+        StackFrame {
+            callee: callee.clone(),
+            memory,
+            argument: Argument::Query(query),
+            ret: Default::default(),
+        }
+    }
+
+    fn new_transaction(
+        callee: &ContractId,
         memory: MemoryRef,
-        opcode: usize,
-        argument_ptr: usize,
-        argument_len: usize,
-        return_ptr: usize,
-        return_len: usize,
+        transaction: Transaction,
     ) -> Self {
         StackFrame {
-            callee,
+            callee: callee.clone(),
```

Comment:
If we are cloning, then we should receive `callee` by value, and not reference

Reply:
agreed.